--- netqmail-1.05/realrcptto.c.orig	2007-05-26 13:20:04.000000000 +0200
+++ netqmail-1.05/realrcptto.c	2007-05-26 17:29:29.000000000 +0200
@@ -152,6 +152,21 @@
   }
 }
 
+static int hasdeny(b)
+stralloc *b;
+{
+  static stralloc d = {0};
+  struct stat st;
+
+  if (!stralloc_copy(&d,b)) die_nomem();
+  if (!stralloc_cats(&d,".DENY")) die_nomem();
+  if (!stralloc_0(&d)) die_nomem();
+  if (stat(d.s,&st) == 0) return 1;
+  /* Log errors, but let the mail through. */
+  if (errno != error_noent) stat_error(d.s,errno);
+  return 0;
+}
+
 int realrcptto(addr)
 char *addr;
 {
@@ -306,6 +321,7 @@
     if (!stralloc_cats(&qme,"/.qmail")) die_nomem();
     if (!stralloc_cats(&qme,dash)) die_nomem();
     if (!stralloc_cat(&qme,&safeext)) die_nomem();
+    if (hasdeny(&qme)) return denyaddr(addr);
     if (!stralloc_0(&qme)) die_nomem();
     if (stat(qme.s,&st) == 0) return 1;
     if (errno != error_noent) {
@@ -319,6 +335,7 @@
         if (!stralloc_cats(&qme,dash)) die_nomem();
         if (!stralloc_catb(&qme,safeext.s,i)) die_nomem();
         if (!stralloc_cats(&qme,"default")) die_nomem();
+        if (hasdeny(&qme)) return denyaddr(addr);
         if (!stralloc_0(&qme)) die_nomem();
         if (stat(qme.s,&st) == 0) return 1;
         if (errno != error_noent) {
